### ISSUE TITLE
complicated expression fails

### ISSUE THREAD

**Issue by shiozaki** [ValeevGroup/tiledarray#49]
```
MADNESS runtime initialized with 31 threads in the pool and affinity -1 -1 -1
terminate called after throwing an instance of 'TiledArray::Exception'
  what():  TiledArray: exception at /usr/local/include/TiledArray/expressions/variable_list.h(94): assertion failure
Aborted
```
this is what I get from the following complete program. At least it should be written in Wiki that one cannot multiply more than two tensors (if that is the case). Otherwise this is a bug.
```
#include <tiledarray.h> // for TiledArray library

namespace TA = TiledArray;

// Construct a second-order tensor tile that is filled with v
TA::Array<double,2>::value_type
make_tile(TA::Array<double, 2>::range_type& range, const double v) {
  // Construct a tile
  TA::Array<double, 2>::value_type tile(range);
  std::fill(tile.begin(), tile.end(), v); 

  return tile;
}

// Fill array x with value v
void init_array(TA::Array<double, 2>& x, const double v) {
  // Add local tiles to a
  for(TA::Array<double, 4>::iterator it = x.begin(); it != x.end(); ++it) {
    // Construct a tile using a MADNESS task.
    madness::Future<TA::Array<double, 2>::value_type> tile =
        x.get_world().taskq.add(& make_tile, x.trange().make_tile_range(it.ordinal()), v); 

    // Insert the tile into the array
    *it = tile;
  }
}

int main(int argc, char** argv) {
  // Initialize runtime
  madness::World& world = madness::initialize(argc, argv);

  // Construct tile boundary vector
  std::vector<std::size_t> tile_boundaries;
  for(std::size_t i = 0; i <= 16; i += 4)
    tile_boundaries.push_back(i);

  // Construct a set of TiledRange1's
  std::vector<TiledArray::TiledRange1>
    ranges(4, TiledArray::TiledRange1(tile_boundaries.begin(), tile_boundaries.end()));
  std::vector<TiledArray::TiledRange1>
    ranges3(3, TiledArray::TiledRange1(tile_boundaries.begin(), tile_boundaries.end()));


  // Construct the 2D TiledRange
  TA::TiledRange trange(ranges.begin(), ranges.end());
  TA::TiledRange trange3(ranges3.begin(), ranges3.end());

  // Construct array objects.
  TA::Array<double, 4> a(world, trange);
  TA::Array<double, 4> b(world, trange);
  TA::Array<double, 3> c(world, trange3);
  TA::Array<double, 3> d(world, trange3);

  // Initialize a and b.
  a.fill_local(0.0);
  b.fill_local(0.0);
  c.fill_local(0.0);
  d.fill_local(0.0);

  // Compute the contraction c(m,n) = sum_k a(m,k) * b(k,n)
  a("x6,a3,x5,a1") = b("x0,a1,x2,a3") * c("o4,x0,x2") * d("a1,a3,o4") * c("o4,x5,x6");

  // Wait for all the computation to complete before exiting
  world.gop.fence();

  madness::finalize();
  return 0;
}
```

**Comment by justusc** [comment]
Expressions, such as the one you used, are evaluated as a series of binary contractions. They follow the order of operations as specified in the C++ standard.

The problem is that the output of the second contraction yields a zero-order tensor.
```
b("x0,a1,x2,a3") * c("o4,x0,x2") => temp1("a1,a3,o4")
temp1("a1,a3,o4") * d("a1,a3,o4") => temp2("") == ERROR
```

Obviously the error message was not very helpful. I will add check for this type of error and an appropriate message.

**Comment by shiozaki** [comment]
I understand your implementation. However, the library can see the entire line, and therefore, in principle, it should be able to see that a1 and a3 appear in LHS, therefore, should not be summed.

**Comment by justusc** [comment]
I understand your logic, and TA does take into account the result variable list in order to evaluate expressions. However,  expressions like `temp1("a1,a3,o4") * d("a1,a3,o4") => temp2("a1,a3")` are not currently supported because it cannot be mapped to a GEMM operation.

**Comment by shiozaki** [comment]
I thought that for MP2 denominator you support
```
temp2("a1,a3,o4") = temp1("a1,a3,o4") * d("a1,a3,o4");
```
or not? What I wanted to do is
```
 77         i0("o4,a1,a3") = r("x0,a1,x2,a3") * s("o4,x0,x2");
 78         i1("o4,a1,a3") = i0("o4,a1,a3") * d("o4,a1,a3") * 0.5;
 79         t("x2,a3,x0,a1") -= i1("o4,a1,a3") * s("o4,x0,x2");
```
which is identical to my test program above. I believe all the steps are accepted.

**Comment by justusc** [comment]
I must apologize, I must be more tired than I realized. I gave you the wrong analysis for the output. The second contraction should map to an element-wise multiplication, since all variables are identical in the two arguments.

```
b("x0,a1,x2,a3") * c("o4,x0,x2") => temp1("a1,a3,o4") 
temp1("a1,a3,o4") * d("a1,a3,o4") => temp2("a1,a3,o4") == element-wise multiplication
 temp2("a1,a3,o4") * c("o4,x5,x6") => temp3("a1,a3,x5,x6")
temp3("a1,a3,x5,x6") => a("x6,a3,x5,a1")
```

After a second look, I think your expression should work with the element-wise contraction. I will try your code tomorrow, and try to track down the source of the error.
